core/server/master/src/main/java/alluxio/master/file/meta/UfsAbsentPathCache.java

```diff
@@ -0,0 +1,191 @@
+    mCaches.remove(mountId);
+  }
+
+  /**
+   * Adds the given absent path into the cache. This will sequentially walk down the path to find
+   * the first component which does not exist in the ufs.
+   *
+   * @param path the absent path to add to the cache
+   * @throws InvalidPathException if the path is invalid
+   */
+  public void addAbsentPath(AlluxioURI path) throws InvalidPathException {
+    MountTable.Resolution resolution = mMountTable.resolve(path);
+    AlluxioURI ufsUri = resolution.getUri();
+    UnderFileSystem ufs = resolution.getUfs();
+    ConcurrentSkipListSet<String> cache = getCache(resolution.getMountId());
+
+    String[] components = PathUtils.getPathComponents(ufsUri.getPath());
+    if (components.length == 0) {
+      return;
+    }
+
+    // create a ufs uri of the root of the mount point.
+    AlluxioURI uri = mMountTable.getMountInfo(resolution.getMountId()).getUfsUri();
+    int mountPointBase = uri.getDepth();
+
+    // Traverse through the ufs path components, staring from the root, to find the first
+    // non-existing ufs path.
+    for (int i = 0; i < components.length; i++) {
+      if (i > 0 && i <= mountPointBase) {
+        // Do not process components before the base of the mount point.
+        // However, process the first component, since that will be the actual mount point base.
+        continue;
+      }
+      String component = components[i];
+      uri = uri.join(component);
+      String uriPath = uri.getPath();
+      try {
+        if (ufs.exists(uri.toString())) {
+          // This ufs path exists. Remove the cache entry.
+          cache.remove(uriPath);
+        } else {
+          // This is the first ufs path which does not exist. Add it to the cache.
+          cache.add(uriPath);
+
+          // Remove cache entries which has this non-existing directory as a prefix. This is not for
+          // correctness, but to "compress" information.
+          String dirPath = uriPath + "/";
+          Iterator<String> it = cache.tailSet(dirPath).iterator();
+          while (it.hasNext()) {
+            String existingPath = it.next();
+            if (existingPath.startsWith(dirPath)) {
+              // An existing cache entry has the non-existing path as a prefix. Remove the entry,
+              // since the non-existing path ancestor implies the descendant does not exist.
+              it.remove();
+            }
+          }
+          // The first non-existing path was found, so further traversal is unnecessary.
+          break;
+        }
+      } catch (IOException e) {
+        throw new InvalidPathException(ExceptionMessage.PATH_DOES_NOT_EXIST.getMessage(path), e);
+      }
+    }
+  }
+
+  /**
+   * Removes an absent path from the cache.
+   *
+   * @param path the path to remove from the cache
+   * @throws InvalidPathException if the path is invalid
+   */
+  public void removeAbsentPath(AlluxioURI path) throws InvalidPathException {
```

Comment:
optional: this could just be `remove` or `removePath` I think since the class is already called `AbsentPathCache`